### DISCUSSION TITLE
deploy lambda with pmtiles

### DISCUSSION THREAD

**Discussion by MohammadUT** [protomaps/PMTiles#475] [Q&A] [answered]
Hi there, 

I'm working on deploying lambda with pmtiles using the cloudformation .yaml file provide. I have set all configurations as well. All my .pmtiles are stored in S3 bucket, in (e.g., tiles/name.pmtiles). However, I cant directly get access to pmtiles through cloudfront https://CLOUDFRONT_name.net/tiles/jibe_basemap.pmtiles and got this error 'Error: Wrong magic number for PMTiles archive'. Also pmtiles are correctly generated and there is no issue with them. Is the format of provided URL is not correct and if so, what this URL should be look like? 
I have even tried without .pmtiles and still got invalid tile URL when I open the cloudfront URL through browser. Any suggestions on how to solve this would be really appreciated.

**Comment by bdon** [comment, marked as answer]
Your setup looks correct but if you are using lambda you need to access the tiles via ZXY URLs or TileJSON, not as a raw `.pmtiles` file, that's the purpose of the lambda integration. Please follow the docs for your map rendering library on how to add a tiled data source.

**Reply by MohammadUT** [reply]
Thanks for that, I could fix that problem, however got another issue. I could load base map pmtiles successfully, but for another pmtile file which needs to loaded as well, some tiles could successful fetched and loaded, and some of those didn't load. I got the following Error: AJAXError:  (502) for several tiles (e.g.,https://CLOUDFRONT_name.net/tiles/jibe_directional_network/9/462/313.mvt). I have also attached the screenshot of loaded map.

Btw, the request to the Lambda Function URL (https://CLOUDFRONT_name.net/tiles/jibe_directional_network/0/0/0.mvt) succeed. 

Thanks.

![Screenshot 2024-10-18 at 15 29 49](https://github.com/user-attachments/assets/d1b5582f-8357-4ddb-b965-7a356b7da8d5)

**Reply by bdon** [reply]
Thanks for the report, can you please check any generated logs in CloudWatch for errors?

**Reply by bdon** [reply]
Also, is it possible that your generated tile is more than 6 megabytes? That can happen with large datasets, and we should document the limit which is part of AWS Lambda. But for most interactive web maps you really want to generate tiles 1MB or less.

**Reply by MohammadUT** [reply]
Thanks, I noticed some tiles are greater than 6 megabytes. Also, the main .pmtile didn't include all zoom levels like the one mentioned earlier. Therefore, the lambda couldn't fetch the tile for that zoom level and shown as missed tiles. I regenerated the tiles for all zoom levels, and could load the map successfully. Also checked the lambda is working fine and I could see successful logs in cloudwatch as well. However the regenrated tiles dropped significantly the level of detail which is another issue. I'm looking for keeping the most details (e.g., streets) for all zoom layers.

<img width="1042" alt="finall" src="https://github.com/user-attachments/assets/a1c070cf-d746-4136-8953-96ac4d6b097b">

**Reply by bdon** [reply]
I added a note about the 6MB limit here: https://docs.protomaps.com/deploy/aws#installation

Yes, I think the necessary step is to optimize your tileset, possibly by dropping as many properties from features as possible.
